### ISSUE TITLE
vmalert: long query execution in annotations blocks API

### ISSUE THREAD

**Issue by dmitryk-dk** [VictoriaMetrics/VictoriaMetrics#6079]
### Describe the bug

If you are using the `query` template function in the alerting rule annotations and this `query` executes for a long time (let's say 1 minute), the other vmalert endpoint does not return a response.

### To Reproduce

create alerting group like in the example
```
groups:
  - name: group
    interval: 5s
    eval_offset: 0
    rules:
      - alert: Blocking
        expr: 1 + 1
        annotations:
          summary: '{{ query "1" | first | value }}'

      - alert: Next Alert
        expr: vector(1)
        annotations:
          summary: 'just summary'
```

Prepare nginx config

```
daemon off;
error_log  nginx_error.log warn;
pid        nginx.pid;

events { }

http {


    log_format   main '$remote_addr [$time_local]  $status '
        '"$request" $body_bytes_sent "$http_referer" '
        '"$http_user_agent" "$http_x_forwarded_for"';

    limit_req_zone $binary_remote_addr zone=delay:1m rate=1r/m;

    server {
        listen       12345 default_server;
        access_log   nginx_access.log main;
        root /dev/null;

        location /datasource {
            return 200 '{"status":"success","isPartial":false,"data":{"resultType":"vector","result":[{"metric":{},"value":[1712340706,"2"]}]},"stats":{"seriesFetched": "0","executionTimeMsec":0}}\n';
        }

        location /api/v2/alerts {
            return 200;
        }

        location /delay-second-request {
            limit_req zone=delay burst=1;
            proxy_pass http://localhost:12345/datasource/;
        }
    }
}
```
run nginx
```
nginx -p . -e nginx_error.log -c ./bin/nginx.conf
```

Run vmalert 

and try to curl `/metrics` endpoint

```
curl http://localhost:12346/metrics
```
The request will stack, while `query` from the template will be executed, `curl` will wait for the response.

I think it happens because of this mutex
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/83216e956c86707abf6424018d13e57f66ac393b/app/vmalert/rule/alerting.go#L389

### Version

vmalert v1.99.0

**Comment by hagen1778** [comment]
Looks like vmalert is get blocked here https://github.com/VictoriaMetrics/VictoriaMetrics/blob/83216e956c86707abf6424018d13e57f66ac393b/app/vmalert/rule/alerting.go#L97-L120

When /metrics endpoint is called, metrics package invokes all Gauge functions to collect the data. And in vmalert's case, those functions will try to acquire the lock to read the data. 

At the same time, the same AlertingRule object could have already acquired the lock in [Exec function](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/83216e956c86707abf6424018d13e57f66ac393b/app/vmalert/rule/alerting.go#L389-L390). Here the lock is acquired to update the object fields. And one of these fields is `Annotations` which could take extra time to be updated when contains `query` operator inside (executes arbitrary query configured by user). For the rest of cases, when there is no `query` function in annotations, fields updates takes no time.

One solution here is to make Gauge metrics use atomic values instead of acquiring the lock. This should make the metrics collection always independent.

**Comment by Haleygo** [comment]
`/api/v1/rules`  and  `/api/v1/alerts` would be blocked with the same lock problem as described [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6079#issuecomment-2044975574).

**Comment by hagen1778** [comment]
@dmitryk-dk https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6129 has been merged and will be included into the next release.

**Comment by valyala** [comment]
FYI, this bugfix has been included in [v1.97.4 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.97.4).

**Comment by hagen1778** [comment]
vmalert won't block /api/v1/rules, /api/v1/alerts, /metrics APIs during rules evaluation starting from [v1.101.0](https://docs.victoriametrics.com/changelog/#v11010) release.
